Jenkinsfile

```diff
@@ -0,0 +1,155 @@
+                        url: 'https://github.com/realm/realm-java.git'
+                    ]]
+                ])
+            } else {
+                checkout([
+                    $class: 'GitSCM',
+                    branches: [[name: '*/master']],
+                    doGenerateSubmoduleConfigurations: false,
+                    extensions: [[$class: 'CleanCheckout']],
+                    gitTool: 'native git',
+                    submoduleCfg: [],
+                    userRemoteConfigs: [[url: 'https://github.com/realm/realm-java.git']]
+                ])                
+            }
+
+           stage 'JVM tests'
+           sh 'chmod +x gradlew && ./gradlew assemble check javadoc --stacktrace'
+           storeJunitResults('realm/realm-annotations-processor/build/test-results/TEST-*.xml')
+           if (env.BRANCH_NAME == 'master') {
+               collectAarMetrics()
+           }
+           // TODO: add support for running monkey on the example apps
+           //stash includes: 'examples/*/build/outputs/apk/*debug.apk', name: 'examples'
+
+           dir('examples') {
+              sh 'chmod +x gradlew && ./gradlew check --stacktrace'
+              storeJunitResults('unitTestExample/build/test-results/**/TEST-*.xml')
+           }
+
+           stage 'static code analysis'
+           try {
+                dir('realm') {
+                   sh 'chmod +x gradlew && ./gradlew findbugs pmd checkstyle --stacktrace'
+                }
+           } finally {
+                publishHTML(target: [allowMissing: false, alwaysLinkToLastBuild: false, keepAll: true, reportDir: 'realm/realm-library/build/findbugs', reportFiles: 'findbugs-output.html', reportName: 'Findbugs issues'])
+                publishHTML(target: [allowMissing: false, alwaysLinkToLastBuild: false, keepAll: true, reportDir: 'realm/realm-library/build/reports/pmd', reportFiles: 'pmd.html', reportName: 'PMD Issues'])
+           }
+
+           stage 'build instrumented tests'
+           dir('realm') {
+               sh 'chmod +x gradlew && ./gradlew assembleDebugAndroidTest --stacktrace'
+               dir('realm-library/build/outputs/apk') {
+                  stash name: 'test-apk', includes: 'realm-android-library-debug-androidTest-unaligned.apk'
+               }
+           }
+        }
+
+        node('android-hub') {
+            stage 'run instrumented tests'
+            unstash 'test-apk'
+            sh 'adb devices'
+            sh 'adb devices | grep -v List | grep -v ^$ | awk \'{print $1}\' | parallel \'adb -s {} uninstall io.realm.test; adb -s {} install realm-android-library-debug-androidTest-unaligned.apk; adb -s {} shell am instrument -w -r io.realm.test/android.support.test.runner.AndroidJUnitRunner > test_result_{}.txt; java -jar /opt/log-converter.jar test_result_{}.txt\''
```

Comment:
Also, this should probably be more robust against the case where there is no device ID returned by `adb devices`.